### PR TITLE
executor: fix update global index return duplicate key error

### LINKED ISSUE #42312: executor: for global index update partition keys but returns "Duplicate Key" error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```SQL
drop table if exists test_global
create table test_global ( a int, b int, c int)
	partition by range( a ) (
		partition p1 values less than (10),
		partition p2 values less than (20),
		partition p3 values less than (30)
	);
alter table test_global add unique index idx_b (b);
insert into test_global values (1, 1, 1), (8, 8, 8), (11, 11, 11), (12, 12, 12);
update test_global set a = 2 where a = 11;
```

### 2. What did you expect to see? (Required)
update successfully.

### 3. What did you see instead (Required)
Duplicate Key error.
image

### 4. What is your TiDB version? (Required)
image

### PR DESCRIPTION
**Pull request by L-maple**
### What problem does this PR solve?

Issue Number: close #42312

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: table/tables/index.go [hunk 1]
```diff
@@ -400,6 +410,23 @@ func (c *index) Delete(sc *stmtctx.StatementContext, txn kv.Transaction, indexed
 		}
 		tempValElem := tablecodec.TempIndexValueElem{Handle: h, KeyVer: tempKeyVer, Delete: true, Distinct: distinct}
 
+		// If index is global, decode the pid from value and compare with c.physicalID.
+		// Only when pid in value equals to c.physicalID, the key can be deleted.
+		if c.idxInfo.Global {
+			value, err := txn.GetMemBuffer().Get(context.Background(), key)
+			segs := tablecodec.SplitIndexValue(value)
```
[the remark this answers is not in the document]
Comment (L-maple) [reply table/tables/index.go]: resolved

**Comment by CbcWestwolf** [comment]
To pass the `check_dev`, please run `make errdoc` and generate a new `errors.toml`

**Comment by L-maple** [comment]
> To pass the `check_dev`, please run `make errdoc` and generate a new `errors.toml`

resolved

File: errno/errcode.go [hunk 2]
```diff
@@ -924,6 +924,7 @@ const (
 	ErUserAccessDeniedForUserAccountBlockedByPasswordLock    = 3955
 	ErrJSONInBooleanContext                                  = 3986
 	ErrTableWithoutPrimaryKey                                = 3750
+	ErrReorgPartitionSetForGlobalIndex                       = 3987
```
[the remark this answers is not in the document]
Comment (L-maple) [reply errno/errcode.go]: resolved

File: table/tables/partition.go [hunk 3]
```diff
@@ -1590,6 +1599,14 @@ func partitionedTableUpdateRecord(gctx context.Context, ctx sessionctx.Context,
 			return errors.WithStack(table.ErrRowDoesNotMatchGivenPartitionSet)
 		}
 	}
+	// TODO(L-Maple): Handle global index.
```
Comment (L-maple) [inline table/tables/partition.go:1602]: @mjonss onDropPartition is affected by onReorganizePartition. 
1.When a table has a global index, `DroppingDefinitions` field is also set. Maybe we ignore it in the code logic.
2.For example, pt has p0,p1,p2 and p0 is in dropPartition reorg. If we write to p0, `to` is the pid of p1. That's actually wrong.

File: table/tables/partition.go [hunk 4]
```diff
@@ -1687,6 +1687,15 @@ func partitionedTableUpdateRecord(gctx context.Context, ctx sessionctx.Context,
 	return nil
 }
 
+func hasGlobalIndex(tblInfo *model.TableInfo) bool {
```
Comment (Defined2014) [inline table/tables/partition.go:1690]: Does this function be called?

Comment (L-maple) [reply table/tables/partition.go]: resolved. this function is no longer used, and I removed it.

**Review by Defined2014 — APPROVED** [review]
LGTM for this PR, Not sure should we concerned about rollback for this situation. /cc @tiancaiamao 
https://github.com/pingcap/tidb/blob/fc166ffae8576ac3efb2f5abac848d5edbaf2657/table/tables/partition.go#L1604-L1609

File: ddl/db_partition_test.go [hunk 5]
```diff
@@ -1952,6 +1953,106 @@ func TestDropPartitionWithGlobalIndex(t *testing.T) {
 	require.Equal(t, 2, cnt)
 }
 
+func TestGlobalIndexInsertInDropPartition(t *testing.T) {
```
Comment (mjonss) [inline ddl/db_partition_test.go:1956]: Is this test suppose to hit the change/bug or just there for completeness?

File: ddl/db_partition_test.go [hunk 6]
```diff
@@ -1952,6 +1953,106 @@ func TestDropPartitionWithGlobalIndex(t *testing.T) {
...
+	);`)
+	tk.MustExec("alter table test_global add unique index idx_b (b);")
+	tk.MustExec("insert into test_global values (1, 1, 1), (8, 8, 8), (11, 11, 11), (12, 12, 12);")
+	tk.MustExec("update test_global set a = 2 where a = 11")
+	tk.MustExec("update test_global set a = 13 where a = 12")
+	tk.MustExec("analyze table test_global")
+	tk.MustQuery("select * from test_global use index(idx_b) order by a").Check(testkit.Rows("1 1 1", "2 11 11", "8 8 8", "13 12 12"))
+}
+
+func TestGlobalIndexUpdateInDropPartition(t *testing.T) {
```
Comment (mjonss) [inline ddl/db_partition_test.go:2018]: Same here, is this test suppose to hit the change/bug or just there for completeness?

**Review by mjonss — APPROVED** [review]
LGTM

**Comment by L-maple** [comment]
just for complement. there is a ut for drop partition with global index，but no ut exist for write when ddl happens.



---- Replied Message ----
| From | Mattias ***@***.***> |
| Date | 03/22/2023 23:22 |
| To | ***@***.***> |
| Cc | Changjie ***@***.***>***@***.***> |
| Subject | Re: [pingcap/tidb] executor: fix update global index return duplicate key error (PR #42311) |

@mjonss approved this pull request.

LGTM

In ddl/db_partition_test.go:

> @@ -1952,6 +1953,106 @@ func TestDropPartitionWithGlobalIndex(t *testing.T) {
 	require.Equal(t, 2, cnt)
 }
 
+func TestGlobalIndexInsertInDropPartition(t *testing.T) {


Is this test suppose to hit the change/bug or just there for completeness?

In ddl/db_partition_test.go:

> +	tk.MustExec("drop table if exists test_global")
+	tk.MustExec(`create table test_global ( a int, b int, c int)
+	partition by range( a ) (
+		partition p1 values less than (10),
+		partition p2 values less than (20),
+		partition p3 values less than (30)
+	);`)
+	tk.MustExec("alter table test_global add unique index idx_b (b);")
+	tk.MustExec("insert into test_global values (1, 1, 1), (8, 8, 8), (11, 11, 11), (12, 12, 12);")
+	tk.MustExec("update test_global set a = 2 where a = 11")
+	tk.MustExec("update test_global set a = 13 where a = 12")
+	tk.MustExec("analyze table test_global")
+	tk.MustQuery("select * from test_global use index(idx_b) order by a").Check(testkit.Rows("1 1 1", "2 11 11", "8 8 8", "13 12 12"))
+}
+
+func TestGlobalIndexUpdateInDropPartition(t *testing.T) {


Same here, is this test suppose to hit the change/bug or just there for completeness?

File: table/tables/index.go [hunk 7]
```diff
@@ -400,6 +410,23 @@ func (c *index) Delete(sc *stmtctx.StatementContext, txn kv.Transaction, indexed
...
+					_, pid, err := codec.DecodeInt(segs.PartitionID)
+					if err != nil {
+						return err
+					}
+					if pid != c.phyTblID {
+						continue
+					}
+				}
+			}
+		}
```
Comment (mjonss) [inline table/tables/index.go:428]: I was a bit curious and I wonder if we should change the way we add the indexes [here](https://github.com/pingcap/tidb/blob/6897c1abb8adb782a474ceaeda005f6cddc10013/table/tables/tables.go#L186-L198)?
I mean if it is a global index, why add it with the physical table ids of the partitions?

**Pull request merged**